Improve tutorials

cgohlke opened the pull request:
## Description

This PR proposes the following improvements to the tutorials:

- Move tutorials focused on using specific modules, classes, and functions into a "Programming interface" section.
- Create an "Applications" section for future problem-oriented how-to guides for advanced applications.
- Expand the introduction tutorial:
  - State that "using the PhasorPy library requires familiarity with the phasor approach, image processing, array programming, and Python."
  - Specify C compilers required to install PhasorPy from source.
  - Define "phasor coordinates".
  - Make aware of undefined/NaN values.
  - Show that signals can be reconstructed from all harmonics.
  - Add reason why calibration is necessary.
  - Show that calibration can be undone.
  - Plot more intermediate results for calibration and filtering.
  - ~~Add conversion to apparent single lifetimes~~.
  - Add section about spectral phasors.

TODO: 

- ~~Add helper function for plotting images to reduce matplotlib code in tutorials~~.
- ~~Replace FLUTe dataset with PTU files~~.

## Release note

Summarize the changes in the code block below to be included in the
[release notes](https://www.phasorpy.org/stable/release.html):

```release-note
Improve tutorials
```

## Checklist

- [ ] The pull request title, summary, and description are concise.
- [ ] Related issues are linked in the description.
- [ ] New dependencies are explained.
- [ ] The source code and documentation can be distributed under the [MIT license](https://www.phasorpy.org/stable/license.html).
- [ ] The source code adheres to [code standards](https://www.phasorpy.org/stable/contributing.html#code-standards).
- [ ] New classes, functions, and features are thoroughly [tested](https://www.phasorpy.org/stable/contributing.html#tests).
- [ ] New, user-facing classes, functions, and features are [documented](https://www.phasorpy.org/stable/contributing.html#documentation).
- [ ] New features are covered in tutorials.
- [ ] No files other than source code, documentation, and project settings are added to the repository.

cgohlke commented:
Hi @bruno-pannunzio : although I labelled this work in progress, I would appreciate your review. Specifically, is this level of detail appropriate for an introductory tutorial? Also, I think it is about time to replace the TIFF dataset from FLUTe with our own data before the first release of the library. Preferably in form of PTU files to demonstrate the ``phasorpy.io`` module and metadata handling.

bruno-pannunzio commented:
Hi @cgohlke! I really like the suggested changes to the tutorials! It makes it much more clear and more direct to select one depending on the needs. Also the addition of the "Applications" sections is great, this can be useful to show the full functionalities of the library without crowding the API tutorials. I think we can work after in useful tutorials for the Applications section. 

Regarding the introduction tutorial, I think all are excellent additions. 

> is this level of detail appropriate for an introductory tutorial

I believe that in general it's okay the level of detail. I would suggest just moving the apparent lifetime section to another tutorial (maybe to an application tutorial?). I think the introduction tutorial is enough if we show the most basic workflow which I think is reached when plotting. I think further analyses can be discussed in more detail in other tutorials. Also, the introduction tutorial is missing hyperspectral images, which would be good to show since it's also an important aspect of the library, and at the moment the introduction tutorial is biased towards FLIM. Maybe we can use the paramecium example used in the cursors tutorial and show the basic workflow until plotting, which shouldn't extend too much the tutorial. 

> Also, I think it is about time to replace the TIFF dataset from FLUTe with our own data before the first release of the library. Preferably in form of PTU files to demonstrate the `phasorpy.io` module and metadata handling.

Definitely! This is long due. The problem we have with this is that our instruments only generate FBD files. We should consider reaching someone open to provide us some nice PTU files. Maybe @lmalacrida can help with this.

bruno-pannunzio commented on tutorials/api/phasorpy_cursors.py:
I was also going to propose some minor changes to the cursors tutorial in a separate PR, but since this one is related to it, maybe you can add the changes here. The only modifications were to replace the `threshold = mean > 1` line with the `phasor_threshold` function. This was going to be the main proposed change:

`mean, real, imag = phasor_threshold(*phasor_from_signal(signal, axis=0), mean_min=1)`

And also replace where `real[threshold]` and `imag[threshold]` with just `real` and `imag`.

cgohlke replied:
Done

cgohlke commented:
Thanks for the review!

> I think we can work after in useful tutorials for the Applications section.

I'd like to add (not in this PR) a tutorial demonstrating out-of-core processing with dask on one of the tiled datasets that @schutyb recently shared. Let's discuss in the next meeting.

> I would suggest just moving the apparent lifetime section to another tutorial (maybe to an application tutorial?). I think the introduction tutorial is enough if we show the most basic workflow which I think is reached when plotting.

I agree, that section stands out. On the other hand, I wanted the introduction tutorial to demonstrate all aspects of the library, which includes converting phasor coordinates, and (currently still missing) cursors and components. 

> Also, the introduction tutorial is missing hyperspectral images, which would be good to show since it's also an important aspect of the library, and at the moment the introduction tutorial is biased towards FLIM.

True. On the other hand, of the almost 60 functions and classes in the library so far, more than half are specific to FLIM, while only two (?) functions are specific to spectral imaging.

Let's dedicate one section at the end of the introduction tutorial to "Spectral phasors", basically a shortened version of the cursors tutorial. OK?

> The problem we have with this is that our instruments only generate FBD files.

FDB is certainly more appropriate than TIFF for the tutorial, but FBD is such a mess of a format that I prefer to not cover it in the main tutorial.

bruno-pannunzio commented:
> Let's dedicate one section at the end of the introduction tutorial to "Spectral phasors", basically a shortened version of the cursors tutorial. OK?

Great!

> FDB is certainly more appropriate than TIFF for the tutorial, but FBD is such a mess of a format that I prefer to not cover it in the main tutorial.

Agree. PTU is more adequate for the tutorial. Let's discuss next meeting candidates to let us use some of their files.

cgohlke commented:
I added the section about spectral phasors and removed apparent lifetimes. In principle, this is ready for review and could be merged. Switching to PTU files and adding a tutorial about apparent lifetimes (which would require a helper function for plotting images) is better done in separate PRs.

bruno-pannunzio reviewed (approved):
Hi @cgohlke, I think now the introduction tutorial is great. I can work in a separate PR in the sections that were left to do (cursors and components) if you want, so that this can be merged and available.
